Fix #1253 and disallow illegal bodies on HTTP Response

lbguilherme opened the pull request:
As per [HTTP Spec, Section 7.2](http://www.w3.org/Protocols/HTTP/1.0/draft-ietf-http-spec.html#Entity-Body):

> For response messages, whether or not an entity body is included with a message is dependent on both the request method and the response code. All responses to the HEAD request method must not include a body, even though the presence of entity header fields may lead one to believe they do. All 1xx (informational), 204 (no content), and 304 (not modified) responses must not include a body. All other responses must include an entity body or a Content-Length header field defined with a value of zero (0).

Ensuring that a response to a HEAD request has no body is still user's responsibility.
The fix for #1253 is the last sentence.

jhass commented:
I think that's an excellent opportunity for some specs that assure that the behaviour stays that way ;)

lbguilherme commented:
I have reworked everything here:

1. When no `Content-length` is provided for a response, the client must read it until the socket is closed (eof). I have done this by implementing `eof?` on the relevant classes and by creating a `HTTP::UntilEofContent` class.
2. There was a confusion around `@body` and `@body_io`. Only one were used at a time and there was no clear way to know which from a plain `Response` object. Some methods simply assumed `@body`. I have refactored it so that `@body` is always an `IO`, I believe this is cleaner and provides all features of the previous organization.
3. Status 100. 101, 204 and 304 should never have a body. This is enforced by an `ArgumentError` on the constructor and a `InvalidResponse` on `from_io`.
4. When creating a response with any other status code, the default body must be an empty body, with `Content-length` set to zero.

jhass commented:
How about having only the IO internally, but provide convenience getters? `getter body_io` and `def body; body_io.read` end`?

lbguilherme commented:
What about this scenario:

```crystal
a = response.body_io.read(4) # This may be inside some other method.
b = response.body # User then does this unaware of the previous read.
```

Is that ok for `response.body` to sometimes not return the full body?

jhass commented:
Mh, good point, that would probably be unexpected. I'd still really like a concise API for the common case though. So the two and a half things I see we could do in that case:

* Assert that `.pos == 0`, raise `ArgumentError` if not, explaining that you shouldn't mix the calls.
* Or `seek` back to the start, returning the full body always.
  * Or maybe even store `pos` and seek back to it after reading the full body.

lbguilherme commented:
`seek` is not really an option with sockets, but raising an error seems reasonable.

lbguilherme commented:
@jhass Done

jhass commented on spec/std/http/response_spec.cr:
```diff
@@ -78,6 +80,49 @@ module HTTP
+        response.body
+      rescue ex
+        error = ex
+      end
```
We got an `expect_raises` macro for these cases, `expect_raises ArgumentError do`

lbguilherme replied:
Oh, this is far better!

jhass commented:
Looking good otherwise, thank you!

asterite commented:
@lbguilherme Thank you so much!

I started reading this pull request yesterday and thought it was good, regarding the check for body emptiness, and setting a content-length. Now it has many more things which I'm not sure about. Please try to keep pull requests small and focused, otherwise it's impossible for us to accept them if we find just a tiny thing that we don't agree with.

When we added support for HTTP::Client body streaming, we though of two uses cases:

```crystal
# Doesn't stream, Response has a String body, body_io is nil
HTTP::Client.get "..."

# Streams, Response has empty body, body_io is an IO
HTTP::Client.get("...") do |response|
  response.body_io.read
end
# after the block, body_io is closed
```

We also though about making `body` read `body_io` if it's nil. We think that those two use cases won't overlap, and if you use a block you know you must read from `body_io`, because you expect that to be in "stream mode".

But the normal case is to not stream (maybe), so making `body` always be an IO is more tedious to work with, and then you have to remember to consume the IO if you'll make a next request using the same client.

Also, we don't think `IO#eof?` makes sense, because in the case of a socket or pipe you can't determine this, and just blocking here seems wrong (on unintuitive).

lbguilherme commented:
> Please try to keep pull requests small and focused, otherwise it's impossible for us to accept them if we find just a tiny thing that we don't agree with.

Oops, sorry. I ended with a chain of dependencies between the commits so I couldn't separate them very well. I'll reorganize it.

> We also though about making body read body_io if it's nil. We think that those two use cases won't overlap, and if you use a block you know you must read from body_io, because you expect that to be in "stream mode".

The two interfaces are still available, but with an unified syntax. Imagine a scenario where you receive a `Response` from some other third-party class. Is is streamed or not? And if it not, then you can't get a stream from it. You must always rely on having two interfaces for getting responses, just like `get` does. The proposal is to make `response.body` lazy. If you need it, it will read the body and give back a String. And why put `body_io` on the trash if you had have it anyway for reading the String `body`? I think it is better to always allow both. In that case, `HTTP::Client.get("...") do |response| ... end` would be removed too. What do you think?

> But the normal case is to not stream (maybe), so making body always be an IO is more tedious to work with, and then you have to remember to consume the IO if you'll make a next request using the same client.

Last last commit adds `body` as an alias for `body_io`, so this is kind of fixed.

> Also, we don't think IO#eof? makes sense, because in the case of a socket or pipe you can't determine this, and just blocking here seems wrong (on unintuitive).

I would expect `IO#eof?` on a socket/pipe to return true only if it is closed, this makes sense to me, but maybe is not very intuitive. Anyway, it is not really needed here and I'll remove the need for it. We can live without eof.

---

I'll refactor this into other(s) pull request(s) and when ready, I'll close this one to keep things clear. Thanks!

asterite commented:
The problem is that the body's IO must be consumed before executing the next request. Try this:

```crystal
require "http/client"

client = HTTP::Client.new "www.google.com"
response1 = client.get "/"
response2 = client.get "/"
puts response2.body
``` 

I get `invalid Int32: http-equiv="content-type" (ArgumentError)`.

However, I think that one solution to this would be to have HTTP::Client store the last response. Before executing the next response it would consume the last one. The above would work (I just tried it) and we could always work with an IO.

/cc @waj

lbguilherme commented:
> The problem is that the body's IO must be consumed before executing the next request. 

You got me. Now I'm not really sure I want the interface to be modified.

---

#1315 includes the meaningful fixes from this pull, with no interface change.

lbguilherme commented:
I believe this can be closed.

asterite commented:
@lbguilherme Sure
